### PR TITLE
GCP function deployment via workflows

### LINKED ISSUE #5: Automate or Script deployment of functions
It would be good to have some scripts that can generate the needed files/scripts for deploying functions for the various hyperparameters

- [ ] Setting up the data needed by each function
- [ ] Writing the `main.py` and `requirements.txt` files
- [ ] deploy using `gcloud functions deploy` command

### PR DESCRIPTION
**Pull request by falquaddoomi**
Closes #5 

This PR adds one workflow per function (i.e., one for `convert-ids` and one for `ml`) that redeploys the function whenever anything is merged into `main` that alters the function's code.

The PR includes a helper workflow, `helper-deploy-func.yaml`, which does the following:
1. installs the gcloud tools
2. authenticates to GCP as the `cloud-function-deployer` service account, via a service account credentials file stored as a secret in this repo
3. pulls the data archive associated with the current function from a GCS bucket, unless it's already in the GitHub actions cached
4. extracts the data archive, caches the resulting data under a hash of the data archive
5. deploys the function

Each function mentioned above uses the function deployer helper workflow to minimize code duplication.

### CODE REVIEW

File: .github/workflows/deploy-func-convert_ids.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  deploy-convert-ids-func:
+    uses: ./.github/workflows/helper-deploy-func.yaml
+    with:
+      func-name: "gpz-convert-ids"
+      func-src-dir: "functions/convert_ids/convert_ids_deploy"
+      func-entrypoint: "convert_ids"
+      func-memory-mb: 1024
+      func-data-gcs-url: "gs://geneplexus-func-data/convert-ids/convert-ids_data.tar.gz"
+    permissions:
+      contents: 'read'
```
Comment (vincerubinetti) [inline .github/workflows/deploy-func-convert_ids.yaml:20]: Is this necessary? I think the github_token gets this by default.

File: .github/workflows/deploy-func-ml.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    paths:
+      - 'functions/ml/ml_deploy/**'
+
+jobs:
+  deploy-ml-func:
+    uses: ./.github/workflows/helper-deploy-func.yaml
+    with:
+      func-name: "gpz-ml"
+      func-src-dir: "functions/ml/ml_deploy"
+      func-entrypoint: "run_pipeline"
```
Comment (vincerubinetti) [inline .github/workflows/deploy-func-ml.yaml:15]: I renamed this function `ml` in #26 to be consistent.

File: .github/workflows/helper-deploy-func.yaml [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+          # service_account: 'cloud-function-deployer@gap-som-dbmi-geneplx-app-p0n.iam.gserviceaccount.com'
+          credentials_json: ${{ secrets.JSON_GCLOUD_SERVICE_ACCOUNT_JSON }}
+
+      - name: 'Set up Cloud SDK'
+        uses: 'google-github-actions/setup-gcloud@v2'
+        with:
+          version: '>= 363.0.0'
+
+      - name: Get hash of the data archive for this function
+        id: get-data-hash
```
Comment (vincerubinetti) [inline .github/workflows/helper-deploy-func.yaml:70]: I think this is vestigial.

File: .github/workflows/helper-deploy-func.yaml [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        description: GCS URL to the archive containing the data for this function
+      func-memory-mb:
+        required: true
+        type: number
+
+      func-runtime:
+        default: python311
+        type: string
+      func-svc-acct:
+        default: logging-monitoring@gap-som-dbmi-geneplx-app-p0n.iam.gserviceaccount.com
```
Comment (vincerubinetti) [inline .github/workflows/helper-deploy-func.yaml:29]: With these new workflows, do we still need the `ml_deploy.sh` scripts?

**Comment by vincerubinetti** [comment]
I'll try forking this and testing it, short of actually providing the secrets and deploying to GCP.

File: .github/workflows/deploy-func-convert_ids.yaml (hunk 1 above)
Comment (falquaddoomi) [reply .github/workflows/deploy-func-convert_ids.yaml]: Good call; I've removed it, and everything works fine, so it wasn't necessary.

**Review by falquaddoomi — COMMENTED** [review]
Thanks for the comments, @vincerubinetti! FYI, I merged latest `main` into this branch so that the entrypoint you mentioned would be renamed in this PR; apparently I started this PR before that other one was merged.

Also, just FYI, I tested the function deployment and, after a few tweaks, I was able to deploy `gpz-convert-ids` successfully.  Just for future reference, the changes were the following:
- the `cloud-function-deployer` service account ("SA") needs to have `Service Account User` permissions on the SA that runs the function, i.e. `logging-monitoring`
- the GH action I was trying to use apparently isn't compatible with Gen 2 functions, which is strange since they've been GA for 2+ years now...anyway, the regular `gcloud function deploy` command that was being used in those deploy scripts works fine, so I switched over to that.

**Comment by falquaddoomi** [comment]
> Well, couldn't test much with a fake GCP key, but it seems to look as expected.
> 
> Suggestion: I've been adding this to all my workflows now because it's very handy. If a job fails (or even if it doesn't) you can re-run with debug logging enabled, and it'll trigger this action where it'll pause the workflow and allow you to ssh into the runner.
> 
> ```yaml
>       - if: runner.debug == '1'
>         uses: mxschmitt/action-tmate@v3
> ```
> 
> Generally I put it before any critical or likely-to-fail actions.

Neat! I put it before the deploy line in the function deploy helper workflow, since I could imagine us wanting to investigate the filesystem if a deploy failed or if the function was acting strangely.

Kind of a small thing: I'm using the [GitHub Actions extension](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-github-actions) for VS Code, and it flags the line `- if: runner.debug == '1'` as a warning with the message `Context access might be invalid: debug`. Any idea if that's something that can be safely ignored, or if there's maybe alternate syntax that won't show the warning?

**Comment by vincerubinetti** [comment]
Hmm idk what that warning is about but I've never had that action throw an error in any workflow, debug mode or not.

File: .github/workflows/helper-deploy-func.yaml (hunk 3 above)
Comment (vincerubinetti) [reply .github/workflows/helper-deploy-func.yaml:66]: I just meant the `id`, not the whole step. I don't think you're using the `id` anywhere. Completely inconsequential though.

**Pull request merged**